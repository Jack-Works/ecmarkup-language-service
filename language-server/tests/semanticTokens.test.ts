import { expect, it } from 'vitest'
import {
    SemanticToken,
    type SemanticTokenData,
    SemanticTokenModifiers,
    SemanticTokenTypes,
} from '../src/features/semanticTokens.js'
import type { TextDocument } from '../src/lib.js'
import { betterSnapshot, File } from './File.js'

it('highlights AO header', async () => {
    const semanticTokens = new SemanticToken()
    const { document, textDocument, program } = File.basic`
        <emu-clause id="sec-OpenCalc" type="abstract operation">
            <h1>
                OpenCalc (
                    _input_: an ECMAScript language value,
                    _input2_: an ECMAScript language value,
                    [ _input3_: an ECMAScript language value, ]
                ): either a normal completion containing an Object or a throw completion
            </h1>
            <dl class="header">
            </dl>
        </emu-clause>

        <emu-clause id="sec-OpenCalc" type="abstract operation">
            <h1>
                OpenCalc (
                    _input_: an ECMAScript language value,
                    [_input2_]: an ECMAScript language value,
                ): an Object
            </h1>
            <dl class="header">
            </dl>
        </emu-clause>

        <emu-clause id="sec-OpenCalc" type="abstract operation">
            <h1>
                OpenCalc (
                    _input_,
                    [ _input2_ ],
                ): an Object
            </h1>
            <dl class="header">
            </dl>
        </emu-clause>

        <emu-clause id="sec-OpenCalc" type="abstract operation">
            <h1>
                Runtime Semantics: OpenCalc ( ): an Object
            </h1>
            <dl class="header">
            </dl>
        </emu-clause>
    `
    const result = await semanticTokens.tokenize(document, program, { textDocument })
    expect(printTokens(document, result)).toMatchInlineSnapshot(`
      "<emu-clause id="sec-OpenCalc" type="abstract operation">
          <h1>
              OpenCalc (
              ~~~~~~~~~ function
                  _input_: an ECMAScript language value,
                   ~~~~~ variable
                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ type
                  _input2_: an ECMAScript language value,
                   ~~~~~~ variable
                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ type
                  [ _input3_: an ECMAScript language value, ]
                     ~~~~~~ variable
                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ type
              ): either a normal completion containing an Object or a throw completion
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ type
          </h1>
          <dl class="header">
          </dl>
      </emu-clause>

      <emu-clause id="sec-OpenCalc" type="abstract operation">
          <h1>
              OpenCalc (
              ~~~~~~~~~ function
                  _input_: an ECMAScript language value,
                   ~~~~~ variable
                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ type
                  [_input2_]: an ECMAScript language value,
                    ~~~~~~ variable
                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ type
              ): an Object
                 ~~~~~~~~~ type
          </h1>
          <dl class="header">
          </dl>
      </emu-clause>

      <emu-clause id="sec-OpenCalc" type="abstract operation">
          <h1>
              OpenCalc (
              ~~~~~~~~~ function
                  _input_,
                   ~~~~~ variable
                  [ _input2_ ],
                     ~~~~~~ variable
              ): an Object
                 ~~~~~~~~~ type
          </h1>
          <dl class="header">
          </dl>
      </emu-clause>

      <emu-clause id="sec-OpenCalc" type="abstract operation">
          <h1>
              Runtime Semantics: OpenCalc ( ): an Object
              ~~~~~~~~~~~~~~~~~~ comment
                                 ~~~~~~~~~ function
                                               ~~~~~~~~~ type
          </h1>
          <dl class="header">
          </dl>
      </emu-clause>"
    `)
    expect(result).toMatchSnapshot()
})

it('highlights User Code calls in algorithms', async () => {
    const semanticTokens = new SemanticToken()
    const { document, textDocument, program } = File.basic`
        <emu-alg>
            1. Let _x_ be ? Get(_obj_, _propKey_).
        </emu-alg>
    `
    const result = await semanticTokens.tokenize(document, program, { textDocument })
    expect(printTokens(document, result)).toMatchInlineSnapshot(`
      "<emu-alg>
          1. Let _x_ be ? Get(_obj_, _propKey_).
                          ~~~ function, modifiers: mutable
      </emu-alg>"
    `)
})

function printTokens(document: TextDocument, tokens: SemanticTokenData[]) {
    return betterSnapshot(
        document,
        tokens.map(({ length, modifier, offset, tokenType }) => ({
            offset,
            length,
            annotate: ` ${SemanticTokenTypes[tokenType]}${modifier.length ? ', modifiers: ' : ''}${modifier.map((modifier) => SemanticTokenModifiers[modifier]).join(',')}`,
        })),
    )
}
