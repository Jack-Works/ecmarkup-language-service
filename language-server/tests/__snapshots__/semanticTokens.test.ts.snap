// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`highlights AO header 2`] = `
[
  {
    "length": 9,
    "modifier": [],
    "offset": 74,
    "tokenType": 0,
  },
  {
    "length": 5,
    "modifier": [],
    "offset": 98,
    "tokenType": 1,
  },
  {
    "length": 28,
    "modifier": [],
    "offset": 106,
    "tokenType": 3,
  },
  {
    "length": 6,
    "modifier": [],
    "offset": 149,
    "tokenType": 1,
  },
  {
    "length": 28,
    "modifier": [],
    "offset": 158,
    "tokenType": 3,
  },
  {
    "length": 6,
    "modifier": [],
    "offset": 203,
    "tokenType": 1,
  },
  {
    "length": 28,
    "modifier": [],
    "offset": 212,
    "tokenType": 3,
  },
  {
    "length": 69,
    "modifier": [],
    "offset": 255,
    "tokenType": 3,
  },
  {
    "length": 9,
    "modifier": [],
    "offset": 458,
    "tokenType": 0,
  },
  {
    "length": 5,
    "modifier": [],
    "offset": 482,
    "tokenType": 1,
  },
  {
    "length": 28,
    "modifier": [],
    "offset": 490,
    "tokenType": 3,
  },
  {
    "length": 6,
    "modifier": [],
    "offset": 534,
    "tokenType": 1,
  },
  {
    "length": 28,
    "modifier": [],
    "offset": 544,
    "tokenType": 3,
  },
  {
    "length": 9,
    "modifier": [],
    "offset": 585,
    "tokenType": 3,
  },
  {
    "length": 9,
    "modifier": [],
    "offset": 728,
    "tokenType": 0,
  },
  {
    "length": 5,
    "modifier": [],
    "offset": 752,
    "tokenType": 1,
  },
  {
    "length": 6,
    "modifier": [],
    "offset": 775,
    "tokenType": 1,
  },
  {
    "length": 9,
    "modifier": [],
    "offset": 797,
    "tokenType": 3,
  },
  {
    "length": 18,
    "modifier": [],
    "offset": 940,
    "tokenType": 2,
  },
  {
    "length": 9,
    "modifier": [],
    "offset": 959,
    "tokenType": 0,
  },
  {
    "length": 9,
    "modifier": [],
    "offset": 973,
    "tokenType": 3,
  },
]
`;
